storage/filesystem/object.go

```diff
@@ -294,25 +298,29 @@ func newPackfileIter(f billy.File, t plumbing.ObjectType, seen map[plumbing.Hash
 }
 
 func (iter *packfileIter) Next() (plumbing.EncodedObject, error) {
-	if iter.position >= iter.total {
-		return nil, io.EOF
-	}
+	for {
+		if iter.position >= iter.total {
+			return nil, io.EOF
+		}
 
-	obj, err := iter.d.DecodeObject()
-	if err != nil {
-		return nil, err
-	}
+		obj, err := iter.d.DecodeObject()
+		if err != nil {
+			return nil, err
+		}
 
-	iter.position++
-	if iter.seen[obj.Hash()] {
-		return iter.Next()
-	}
+		iter.position++
+		if obj != nil {
```

Comment:
This is what I meant before when commenting on not giving extra work to our users. This method is unnecasary complex due to DecodeObject returning ambigous results. Please change that and simplify this.

Besides this: this `if` should negated to avoid extra logic an extra indentation:
```
if obj == nil {
    continue
}
```